File: spring-boot-rag-ollama/src/main/java/com/example/rag/service/OllamaService.java
package com.example.rag.service;

import com.example.rag.dto.OllamaDtos.*;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.Comparator;
import java.util.List;
import java.util.Locale;

@Service
public class OllamaService {

    private final RestClient restClient;

    public OllamaService(@Value("${ollama.base-url}") String ollamaBaseUrl) {
        this.restClient = RestClient.builder()
                .baseUrl(ollamaBaseUrl)
                .build();
    }

    public List<ModelInfo> getAvailableModels() {
        TagsResponse response = restClient.get()
                .uri("/api/tags")
                .retrieve()
                .body(TagsResponse.class);

        if (response == null || response.models() == null) {
            return List.of();
        }

        return response.models();
    }

    public String detectChatModelName() {
        return getAvailableModels().stream()
                .filter(model -> !isEmbeddingModel(modelName(model)))
                .sorted(Comparator.comparing(ModelInfo::modified_at, Comparator.nullsLast(Comparator.reverseOrder())))
                .map(this::modelName)
                .filter(name -> name != null && !name.isBlank())
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("Kein Chat-Modell in Ollama gefunden."));
    }

    public String detectEmbeddingModelName() {
        return getAvailableModels().stream()
                .filter(model -> isEmbeddingModel(modelName(model)))
                .map(this::modelName)
                .filter(name -> name != null && !name.isBlank())
                .findFirst()
                .orElseThrow(() -> new IllegalStateException(
                        "Kein Embedding-Modell in Ollama gefunden. Bitte z. B. `ollama pull nomic-embed-text` oder `ollama pull embeddinggemma` ausführen."
                ));
    }

    public List<Double> embed(String text) {
        String embeddingModel = detectEmbeddingModelName();
        EmbedRequest request = new EmbedRequest(embeddingModel, text);

        EmbedResponse response = restClient.post()
                .uri("/api/embed")
                .body(request)
                .retrieve()
                .body(EmbedResponse.class);

        if (response == null || response.embeddings() == null || response.embeddings().isEmpty()) {
            throw new IllegalStateException("Ollama hat keine Embeddings zurückgegeben.");
        }

        return response.embeddings().get(0);
    }

    public String generateAnswer(String context, String question) {
        String chatModel = detectChatModelName();

        String prompt = """
                Du bist ein lokaler RAG-Assistent.
                Beantworte die Frage ausschließlich anhand des bereitgestellten Kontexts.
                Antworte auf Deutsch.
                Wenn die Antwort nicht im Kontext steht, antworte exakt:
                Diese Information steht nicht im Dokument.

                KONTEXT:
                %s

                FRAGE:
                %s

                ANTWORT:
                """.formatted(context, question);

        GenerateRequest request = new GenerateRequest(chatModel, prompt, false);

        GenerateResponse response = restClient.post()
                .uri("/api/generate")
                .body(request)
                .retrieve()
                .body(GenerateResponse.class);

        if (response == null || response.response() == null) {
            throw new IllegalStateException("Ollama hat keine Antwort zurückgegeben.");
        }

        return response.response().trim();
    }

    private boolean isEmbeddingModel(String name) {
        if (name == null) {
            return false;
        }

        String lower = name.toLowerCase(Locale.ROOT);
        return lower.contains("embed")
                || lower.contains("embedding")
                || lower.contains("minilm")
                || lower.contains("bge");
    }

    private String modelName(ModelInfo model) {
        if (model == null) {
            return null;
        }
        if (model.name() != null && !model.name().isBlank()) {
            return model.name();
        }
        return model.model();
    }
}

File: rag-demo-py/ollamaservice.py
import  requests
from config import OLLAMA_BASE_URL


class OllamaService:
    def __init__(self):
        self.base_url = OLLAMA_BASE_URL

    def get_models(self) -> list[dict]:
        response = requests.get(f"{self.base_url}/api/tags", timeout=10)
        response.raise_for_status()
        return response.json().get("models", [])

    def detect_embedding_model(self) -> str:
        models = self.get_models()

        for model in models:
            name = model.get("name", "").lower()
            if "embed" in name or "embedding" in name or "bge" in name or "minilm" in name:
                return model["name"]

        raise RuntimeError(
            "Kein Embedding-Modell gefunden. "
            "Bitte z. B. ausführen: ollama pull nomic-embed-text"
        )

    def detect_chat_model(self) -> str:
        models = self.get_models()

        chat_models = []
        for model in models:
            name = model.get("name", "").lower()
            if not any(term in name for term in ["embed", "embedding", "bge", "minilm"]):
                chat_models.append(model)

        if not chat_models:
            raise RuntimeError(
                "Kein Chat-Modell gefunden. "
                "Bitte z. B. ausführen: ollama pull llama3.2"
            )

        return chat_models[0]["name"]

    def list_models(self) -> list[str]:
        return [m.get("name", "") for m in self.get_models()]

    def embed(self, text: str, model: str = None) -> list[float]:
        if model is None:
            model = self.detect_embedding_model()

        payload = {
            "model": model,
            "input": text
        }

        response = requests.post(
            f"{self.base_url}/api/embed",
            json=payload,
            timeout=120
        )
        response.raise_for_status()

        data = response.json()
        embeddings = data.get("embeddings")

        if not embeddings:
            raise RuntimeError("Ollama hat keine Embeddings geliefert.")

        return embeddings[0]

    def generate_answer(self, context: str, question: str, model: str = None) -> str:
        if model is None:
            model = self.detect_chat_model()

        prompt = f"""
Du bist ein RAG-Assistent.

Beantworte die Frage ausschließlich auf Basis des bereitgestellten Kontexts.
Wenn die Antwort nicht im Kontext steht, sage:
"Diese Information steht nicht im Dokument."

KONTEXT:
{context}

FRAGE:
{question}

ANTWORT:
"""

        payload = {
            "model": model,
            "prompt": prompt,
            "stream": False
        }

        response = requests.post(
            f"{self.base_url}/api/generate",
            json=payload,
            timeout=180
        )
        response.raise_for_status()

        return response.json().get("response", "").strip()